src/mlpack/tests/kde_test.cpp

```diff
@@ -0,0 +1,621 @@
+ * Tests when an empty query set is given to be evaluated.
+ */
+BOOST_AUTO_TEST_CASE(EmptyQuerySetTest)
+{
+  arma::mat reference = arma::randu(1, 10);
+  arma::mat query;
+  arma::vec estimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.7;
+  const double relError = 0.01;
+
+  // KDE
+  metric::EuclideanDistance metric;
+  GaussianKernel kernel(kernelBandwidth);
+  KDE<metric::EuclideanDistance,
+      arma::mat,
+      kernel::GaussianKernel,
+      tree::KDTree>
+    kde(metric, kernel, relError, 0.0);
+  kde.Train(reference);
+
+  // When evaluating using the query dataset matrix
+  BOOST_REQUIRE_NO_THROW(kde.Evaluate(query, estimations));
+
+  // When evaluating using a query tree
+  typedef KDTree<EuclideanDistance, kde::KDEStat, arma::mat> Tree;
+  std::vector<size_t> oldFromNewQueries;
+  Tree* queryTree = new Tree(query, oldFromNewQueries, 3);
+  BOOST_REQUIRE_NO_THROW(
+    kde.Evaluate(queryTree, oldFromNewQueries, estimations));
+
+  delete queryTree;
+}
+
+/**
+ * Tests serialiation of KDE models.
+ */
+BOOST_AUTO_TEST_CASE(SerializationTest)
+{
+  // Initial KDE model to me serialized.
+  const double relError = 0.25;
+  const double absError = 0.0;
+  arma::mat reference = arma::randu(4, 800);
+  KDE<metric::EuclideanDistance,
+      arma::mat,
+      kernel::GaussianKernel,
+      tree::KDTree>
+    kde(0.25, relError, absError);
+  kde.Train(reference);
+
+  // Get estimations to compare.
+  arma::mat query = arma::randu(4, 100);;
+  arma::vec estimations = arma::vec(query.n_cols, arma::fill::zeros);
+  kde.Evaluate(query, estimations);
+
+  // Initialize serialized objects.
+  KDE<metric::EuclideanDistance,
+      arma::mat,
+      kernel::GaussianKernel,
+      tree::KDTree> kdeXml, kdeText, kdeBinary;
+  SerializeObjectAll(kde, kdeXml, kdeText, kdeBinary);
+
+  // Check everything is correct.
+  BOOST_REQUIRE_CLOSE(kde.RelativeError(), relError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeXml.RelativeError(), relError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeText.RelativeError(), relError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeBinary.RelativeError(), relError, 1e-8);
+
+  BOOST_REQUIRE_CLOSE(kde.AbsoluteError(), absError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeXml.AbsoluteError(), absError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeText.AbsoluteError(), absError, 1e-8);
+  BOOST_REQUIRE_CLOSE(kdeBinary.AbsoluteError(), absError, 1e-8);
+
+  BOOST_REQUIRE_EQUAL(kde.IsTrained(), true);
+  BOOST_REQUIRE_EQUAL(kdeXml.IsTrained(), true);
+  BOOST_REQUIRE_EQUAL(kdeText.IsTrained(), true);
+  BOOST_REQUIRE_EQUAL(kdeBinary.IsTrained(), true);
+
+  // Test if execution gives the same result.
+  arma::vec xmlEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec textEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec binEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+
+  kde.Evaluate(query, xmlEstimations);
+  kde.Evaluate(query, textEstimations);
+  kde.Evaluate(query, binEstimations);
```

Comment:
Shouldn't these be `kdeXml`, `kdeText`, and `kdeBinary`? :)

Reply:
Exactly, thanks for pointing it out :)
Solved in 93c8191f5020acae898325e24bba4838671d0a05